[BUG] `join on` the condition `cast <row> to <boolean>` brings errors on the GPU engines

**Describe the bug**
An error (Cannot convert GpuCast to AST) occurred when `join on` the condition `cast <row> to <boolean>` on the GPU engines.

**Steps/Code to reproduce bug**
```python
import os
import pandas as pd
os.environ['JAVA_HOME'] = "/usr/lib/jvm/java-17-openjdk-amd64"
os.environ['SPARK_HOME'] = "./spark-4.0.1-bin-hadoop3"
os.environ['PYSPARK_SUBMIT_ARGS'] = "--jars ./rapids-4-spark_2.13-25.10.0.jar,./cudf-25.10.0-cuda12.jar --master local[*] pyspark-shell"

import findspark
findspark.init()
from pyspark.sql import SparkSession
from pyspark.conf import SparkConf
from pyspark.sql import functions as F

spark = SparkSession.builder.appName('SparkRAPIDS').config('spark.plugins','com.nvidia.spark.SQLPlugin').config("spark.executor.memory", "8g").config("spark.driver.memory", "8g").getOrCreate()
spark.sparkContext.addPyFile('./rapids-4-spark_2.13-25.10.0.jar')
spark.sparkContext.addPyFile('./cudf-25.10.0-cuda12.jar')
spark.conf.set('spark.rapids.sql.incompatibleOps.enabled', 'true')
spark.conf.set('spark.rapids.sql.format.csv.read.enabled', 'true')
spark.conf.set('spark.rapids.sql.format.csv.enabled', 'true')
spark.conf.set("spark.executor.resource.gpu.amount", "1")
spark.conf.set("spark.task.resource.gpu.amount", "1")
spark.conf.set("spark.rapids.sql.concurrentGpuTasks", "1")
spark.conf.set("spark.rapids.sql.exec.CollectLimitExec", "true")
spark.conf.set('spark.rapids.sql.enabled', 'false')

pd_t0 = pd.DataFrame({
    'c0': [0, 1]
})
df = spark.createDataFrame(pd_t0)
df.createOrReplaceTempView("t0")

pd_t1 = pd.DataFrame({
    'c0': [1],
    'c1': [2]
})
df = spark.createDataFrame(pd_t1)
df.createOrReplaceTempView("t1")

print("CPU:")
cpu_result = spark.table("t1").join(spark.table("t0").alias("t0"), F.col("t0.c0").cast("boolean"), "right_outer").select(F.col("t0.c0"))
cpu_result.show(truncate=False)

spark.conf.set('spark.rapids.sql.enabled', 'true')
print("GPU:")
gpu_result = spark.table("t1").join(spark.table("t0").alias("t0"), F.col("t0.c0").cast("boolean"), "right_outer").select(F.col("t0.c0"))
gpu_result.show(truncate=False)
```
```bash
CPU:
+---+                                                                           
|c0 |
+---+
|0  |
|1  |
+---+

GPU:
25/11/29 10:52:31 WARN GpuOverrides: 
          ! <RDDScanExec> cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.execution.RDDScanExec
            @Expression <AttributeReference> c0#1L could run on GPU
            @Expression <AttributeReference> c1#2L could run on GPU
        ! <RDDScanExec> cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.execution.RDDScanExec
          @Expression <AttributeReference> c0#0L could run on GPU

25/11/29 10:52:31 WARN GpuOverrides: 
    ! <RDDScanExec> cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.execution.RDDScanExec
      @Expression <AttributeReference> c0#1L could run on GPU
      @Expression <AttributeReference> c1#2L could run on GPU

25/11/29 10:52:31 WARN GpuOverrides: 
        ! <RDDScanExec> cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.execution.RDDScanExec
          @Expression <AttributeReference> c0#0L could run on GPU

25/11/29 10:52:31 ERROR Executor: Exception in task 4.0 in stage 5.0 (TID 52)
java.lang.IllegalStateException: Cannot convert GpuCast to AST
        at com.nvidia.spark.rapids.GpuExpression.convertToAst(GpuExpressions.scala:198)
        at com.nvidia.spark.rapids.GpuExpression.convertToAst$(GpuExpressions.scala:197)
        at com.nvidia.spark.rapids.GpuUnaryExpression.convertToAst(GpuExpressions.scala:270)
        at org.apache.spark.sql.rapids.execution.GpuBroadcastNestedLoopJoinExecBase$.nestedLoopJoin(GpuBroadcastNestedLoopJoinExecBase.scala:351)
        at org.apache.spark.sql.rapids.execution.GpuBroadcastNestedLoopJoinExecBase.$anonfun$doConditionalJoin$1(GpuBroadcastNestedLoopJoinExecBase.scala:812)
        at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2(RDD.scala:866)
        at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2$adapted(RDD.scala:866)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:107)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:54)
        at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:171)
        at org.apache.spark.scheduler.Task.run(Task.scala:147)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$5(Executor.scala:647)
        at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:80)
        at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:77)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:99)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:650)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)

25/11/29 10:52:31 ERROR TaskSetManager: Task 12 in stage 5.0 failed 1 times; aborting job
Traceback (most recent call last):
  File "/xxx/bug.py", line 56, in <module>
    gpu_result.show(truncate=False)
  File "/xxx/../library/spark-4.0.1-bin-hadoop3/python/pyspark/sql/classic/dataframe.py", line 285, in show
    print(self._show_string(n, truncate, vertical))
  File "/xxx/../library/spark-4.0.1-bin-hadoop3/python/pyspark/sql/classic/dataframe.py", line 316, in _show_string
    return self._jdf.showString(n, int_truncate, vertical)
  File "../library/spark-4.0.1-bin-hadoop3/python/lib/py4j-0.10.9.9-src.zip/py4j/java_gateway.py", line 1362, in __call__
  File "/xxx/../library/spark-4.0.1-bin-hadoop3/python/pyspark/errors/exceptions/captured.py", line 282, in deco
    return f(*a, **kw)
  File "../library/spark-4.0.1-bin-hadoop3/python/lib/py4j-0.10.9.9-src.zip/py4j/protocol.py", line 327, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling o113.showString.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 12 in stage 5.0 failed 1 times, most recent failure: Lost task 12.0 in stage 5.0 (TID 60) (70e244428b2c executor driver): java.lang.IllegalStateException: Cannot convert GpuCast to AST
        at com.nvidia.spark.rapids.GpuExpression.convertToAst(GpuExpressions.scala:198)
        at com.nvidia.spark.rapids.GpuExpression.convertToAst$(GpuExpressions.scala:197)
        at com.nvidia.spark.rapids.GpuUnaryExpression.convertToAst(GpuExpressions.scala:270)
        at org.apache.spark.sql.rapids.execution.GpuBroadcastNestedLoopJoinExecBase$.nestedLoopJoin(GpuBroadcastNestedLoopJoinExecBase.scala:351)
        at org.apache.spark.sql.rapids.execution.GpuBroadcastNestedLoopJoinExecBase.$anonfun$doConditionalJoin$1(GpuBroadcastNestedLoopJoinExecBase.scala:812)
        at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2(RDD.scala:866)
        at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2$adapted(RDD.scala:866)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:107)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:54)
        at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:171)
        at org.apache.spark.scheduler.Task.run(Task.scala:147)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$5(Executor.scala:647)
        at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:80)
        at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:77)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:99)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:650)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)

Driver stacktrace:
        at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$3(DAGScheduler.scala:2935)
        at scala.Option.getOrElse(Option.scala:201)
        at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2935)
        at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2927)
        at scala.collection.immutable.List.foreach(List.scala:334)
        at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2927)
        at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1295)
        at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1295)
        at scala.Option.foreach(Option.scala:437)
        at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1295)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:3207)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:3141)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:3130)
        at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:50)
        at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:1009)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2484)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2505)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2524)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2549)
        at org.apache.spark.rdd.RDD.$anonfun$collect$1(RDD.scala:1057)
        at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
        at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
        at org.apache.spark.rdd.RDD.withScope(RDD.scala:417)
        at org.apache.spark.rdd.RDD.collect(RDD.scala:1056)
        at org.apache.spark.sql.execution.SparkPlan.executeCollect(SparkPlan.scala:462)
        at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.$anonfun$executeCollect$1(AdaptiveSparkPlanExec.scala:402)
        at org.apache.spark.sql.execution.adaptive.ResultQueryStageExec.$anonfun$doMaterialize$1(QueryStageExec.scala:325)
        at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withThreadLocalCaptured$4(SQLExecution.scala:322)
        at org.apache.spark.sql.execution.SQLExecution$.withSessionTagsApplied(SQLExecution.scala:272)
        at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withThreadLocalCaptured$3(SQLExecution.scala:320)
        at org.apache.spark.JobArtifactSet$.withActiveJobArtifactState(JobArtifactSet.scala:94)
        at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withThreadLocalCaptured$2(SQLExecution.scala:316)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.IllegalStateException: Cannot convert GpuCast to AST
        at com.nvidia.spark.rapids.GpuExpression.convertToAst(GpuExpressions.scala:198)
        at com.nvidia.spark.rapids.GpuExpression.convertToAst$(GpuExpressions.scala:197)
        at com.nvidia.spark.rapids.GpuUnaryExpression.convertToAst(GpuExpressions.scala:270)
        at org.apache.spark.sql.rapids.execution.GpuBroadcastNestedLoopJoinExecBase$.nestedLoopJoin(GpuBroadcastNestedLoopJoinExecBase.scala:351)
        at org.apache.spark.sql.rapids.execution.GpuBroadcastNestedLoopJoinExecBase.$anonfun$doConditionalJoin$1(GpuBroadcastNestedLoopJoinExecBase.scala:812)
        at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2(RDD.scala:866)
        at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2$adapted(RDD.scala:866)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at com.nvidia.spark.rapids.GpuOpTimeTrackingRDD.compute(GpuExec.scala:61)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:107)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:54)
        at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:171)
        at org.apache.spark.scheduler.Task.run(Task.scala:147)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$5(Executor.scala:647)
        at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:80)
        at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:77)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:99)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:650)
        ... 3 more
```

**Expected behavior**
No error occurs when using the GPU engines.
The result is consistent when using the CPU and GPU engines.

**Environment details (please complete the following information)**
 - Environment location: Standalone
 - Spark configuration settings related to the issue:
       java-17-openjdk-amd64
      spark-4.0.1-bin-hadoop3
      rapids-4-spark_2.13-25.10.0
      cudf-25.10.0-cuda12

**Additional context**
Add any other context about the problem here.

--- a/integration_tests/src/main/python/join_test.py
+++ b/integration_tests/src/main/python/join_test.py
@@ -1693,3 +1693,36 @@ def test_join_gatherer_size_estimate_threshold(join_type, threshold, batch_size)
             ("b", RepeatSeqGen(StringGen(pattern="[abc]{1,5}"), length = 5))], nullable=False), 1024, 1024)
         return left_df.join(right_df, left_df.a == right_df.r_a, join_type)
     assert_gpu_and_cpu_are_equal_collect(do_join, conf=join_conf)
+
+
+@ignore_order(local=True)
+@pytest.mark.parametrize("join_type", ["LeftOuter", "RightOuter"], ids=idfn)
+def test_join_degenerate_outer(join_type):
+    # A degenerate join shows up with one side of the join (build or probe) has no columns
+    # This can happen when a condition only deals with a single side of the join.
+    # and the other side has no columns (only rows), because none of the columns
+    # are output. The empty side must be the oposite of the condition side, or there
+    # would be at least one column output by the join (the condition column).
+    if join_type == "LeftOuter":
+        empty_side = "right"
+        condition_side = "left"
+    else:
+        # RightOuter
+        empty_side = "left"
+        condition_side = "right"
+    left_size, right_size = (100, 100)
+    def do_join(spark):
+        left_df = gen_df(spark, [("l_key", RepeatSeqGen([1, 2, 3, 4, None], data_type=IntegerType())), 
+          ("l_value", IntegerGen())], left_size)
+        right_df = gen_df(spark, [("r_key", RepeatSeqGen([1, 2, 3, 4, None], data_type=IntegerType())),
+          ("r_value", IntegerGen())], right_size)
+        if condition_side == "left":
+            cond = [left_df.l_key.cast("boolean")]
+        else:
+            cond = [right_df.r_key.cast("boolean")]
+        temp_df = left_df.join(right_df, cond, join_type)
+        if empty_side == "right":
+            return temp_df.selectExpr("l_key", "l_value")
+        else:
+            return temp_df.selectExpr("r_key", "r_value")
+    assert_gpu_and_cpu_are_equal_collect(do_join)

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/AstUtil.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/AstUtil.scala
@@ -19,8 +19,8 @@ package com.nvidia.spark.rapids
 import scala.collection.mutable
 import scala.collection.mutable.ListBuffer
 
-import org.apache.spark.sql.catalyst.expressions.{AttributeSeq, Expression, ExprId, NamedExpression}
-import org.apache.spark.sql.rapids.catalyst.expressions.{GpuEquivalentExpressions, GpuExpressionEquals}
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, AttributeSeq, Expression, ExprId, NamedExpression}
+import org.apache.spark.sql.rapids.catalyst.expressions.GpuExpressionEquals
 
 
 object AstUtil {
@@ -35,15 +35,15 @@ object AstUtil {
   def canExtractNonAstConditionIfNeed(expr: BaseExprMeta[_], left: Seq[ExprId],
       right: Seq[ExprId]): Boolean = {
     if (!expr.canSelfBeAst) {
-      // It needs to be split since not ast-able. Check itself and childerns to ensure
-      // pushing-down can be made, which doesn't need attributions from both sides.
+      // This expression cannot be AST. We will extract the entire sub-tree (this expression
+      // and all its children). Check if this entire sub-tree only uses one side of the join.
       val exprRef = expr.wrapped.asInstanceOf[Expression]
-      val leftTree = exprRef.references.exists(r => left.contains(r.exprId))
-      val rightTree = exprRef.references.exists(r => right.contains(r.exprId))
-      // Can't extract a condition involving columns from both sides
-      !(rightTree && leftTree)
+      val hasLeft = exprRef.references.exists(r => left.contains(r.exprId))
+      val hasRight = exprRef.references.exists(r => right.contains(r.exprId))
+      // Can extract if it doesn't use both sides (entire sub-tree will be extracted)
+      !(hasLeft && hasRight)
     } else {
-      // Check whether any child contains the case not able to split
+      // This node is AST-able, so recursively check all children
       expr.childExprs.isEmpty || expr.childExprs.forall(
         canExtractNonAstConditionIfNeed(_, left, right))
     }
@@ -52,70 +52,105 @@ object AstUtil {
   /**
    * Extract non-AST functions from join conditions and update the original join condition. Based
    * on the attributes, it decides which side the split condition belongs to. The replaced
-   * condition is wrapped with GpuAlias with new intermediate attributes.
+   * condition is wrapped with GpuAlias with new intermediate attributes. It is assumed that
+   * `canExtractNonAstConditionIfNeed` was already called and returned true.
    *
    * @param condition to be split if needed
    * @param left attributions from left child
    * @param right attributions from right child
-   * @param skipCheck whether skip split-able check
    * @return a tuple of [[Expression]] for remained expressions, List of [[NamedExpression]] for
    *         left child if any, List of [[NamedExpression]] for right child if any
    */
   def extractNonAstFromJoinCond(condition: Option[BaseExprMeta[_]],
-      left: AttributeSeq, right: AttributeSeq, skipCheck: Boolean):
+                                left: AttributeSeq, right: AttributeSeq):
   (Option[Expression], List[NamedExpression], List[NamedExpression]) = {
-    // Choose side with smaller key size. Use expr ID to check the side which project expr
-    // belonging to.
-    val (exprIds, isLeft) = if (left.attrs.size < right.attrs.size) {
-      (left.attrs.map(_.exprId), true)
-    } else {
-      (right.attrs.map(_.exprId), false)
+    
+    condition match {
+      case None => (None, List.empty, List.empty)
+      case Some(cond) =>
+        // List of expression pushing down to left side child
+        val leftExprs: ListBuffer[NamedExpression] = ListBuffer.empty
+        // List of expression pushing down to right side child
+        val rightExprs: ListBuffer[NamedExpression] = ListBuffer.empty
+        // Deduplication map to avoid processing the same expression multiple times
+        val processed = mutable.HashMap.empty[GpuExpressionEquals, Expression]
+
+        val leftExprIds = left.attrs.map(_.exprId).toSet
+        val rightExprIds = right.attrs.map(_.exprId).toSet
+
+        // Extract and convert in a single pass
+        val updatedCondition = extractAndConvert(cond, leftExprIds, rightExprIds,
+          leftExprs, rightExprs, processed)
+
+        (Some(updatedCondition), leftExprs.toList, rightExprs.toList)
     }
-    // List of expression pushing down to left side child
-    val leftExprs: ListBuffer[NamedExpression] = ListBuffer.empty
-    // List of expression pushing down to right side child
-    val rightExprs: ListBuffer[NamedExpression] = ListBuffer.empty
-    // Substitution map used to replace targeted expressions based on semantic equality
-    val substitutionMap = mutable.HashMap.empty[GpuExpressionEquals, Expression]
-
-    // 1st step to construct 1) left expr list; 2) right expr list; 3) substitutionMap
-    // No need to consider common sub-expressions here since project node will use tiered execution
-    condition.foreach(c =>
-      if (skipCheck || canExtractNonAstConditionIfNeed(c, left.attrs.map(_.exprId), right.attrs
-          .map(_.exprId))) {
-        splitNonAstInternal(c, exprIds, leftExprs, rightExprs, substitutionMap, isLeft)
-      })
-
-    // 2nd step to replace expression pushing down to child plans in depth first fashion
-    (condition.map(
-      _.convertToGpu().mapChildren(
-        GpuEquivalentExpressions.replaceWithSemanticCommonRef(_,
-          substitutionMap))), leftExprs.toList, rightExprs.toList)
   }
 
-  private[this] def splitNonAstInternal(condition: BaseExprMeta[_], childAtt: Seq[ExprId],
-      left: ListBuffer[NamedExpression], right: ListBuffer[NamedExpression],
-      substitutionMap: mutable.HashMap[GpuExpressionEquals, Expression], isLeft: Boolean): Unit = {
-    for (child <- condition.childExprs) {
-      if (!child.canSelfBeAst) {
-        val exprRef = child.wrapped.asInstanceOf[Expression]
-        val gpuProj = child.convertToGpu()
-        val alias = substitutionMap.get(GpuExpressionEquals(gpuProj)) match {
-          case Some(_) => None
-          case None =>
-            if (exprRef.references.exists(r => childAtt.contains(r.exprId)) ^ isLeft) {
-              val alias = GpuAlias(gpuProj, s"_agpu_non_ast_r_${left.size}")()
-              right += alias
-              Some(alias)
-            } else {
-              val alias = GpuAlias(gpuProj, s"_agpu_non_ast_l_${left.size}")()
-              left += alias
-              Some(alias)
-            }
-        }
-        alias.foreach(a => substitutionMap.put(GpuExpressionEquals(gpuProj), a.toAttribute))
+  /**
+   * Recursively extract non-AST expressions and convert to GPU in a single pass.
+   * 
+   * @param expr the expression to process
+   * @param leftExprIds expression IDs from the left side
+   * @param rightExprIds expression IDs from the right side
+   * @param leftExprs buffer to collect expressions for left child
+   * @param rightExprs buffer to collect expressions for right child
+   * @param processed map to avoid processing duplicates
+   * @return the converted GPU expression with non-AST sub-trees replaced
+   */
+  private[this] def extractAndConvert(
+      expr: BaseExprMeta[_],
+      leftExprIds: Set[ExprId],
+      rightExprIds: Set[ExprId],
+      leftExprs: ListBuffer[NamedExpression],
+      rightExprs: ListBuffer[NamedExpression],
+      processed: mutable.HashMap[GpuExpressionEquals, Expression]): Expression = {
+    if (!expr.canSelfBeAst) {
+      // This expression cannot be converted to AST - extract the entire sub-tree
+      val exprRef = expr.wrapped.asInstanceOf[Expression]
+      val gpuExpr = expr.convertToGpu()
+      
+      // Check if we've already processed this expression (for deduplication)
+      processed.get(GpuExpressionEquals(gpuExpr)) match {
+        case Some(replacement) => 
+          replacement
+        case None =>
+          // Determine which side this expression belongs to based on its references
+          val referencedExprIds = exprRef.references.map(_.exprId).toSet
+          val referencesLeft = referencedExprIds.exists(leftExprIds.contains)
+          val referencesRight = referencedExprIds.exists(rightExprIds.contains)
+          
+          // Create an alias and add to appropriate side
+          // Note: if it references both sides or neither, it shouldn't happen if 
+          // canExtractNonAstConditionIfNeed passed, but we'll default to left
+          val alias = if (referencesRight && !referencesLeft) {
+            val a = GpuAlias(gpuExpr, s"_agpu_non_ast_r_${rightExprs.size}")()
+            rightExprs += a
+            a
+          } else {
+            val a = GpuAlias(gpuExpr, s"_agpu_non_ast_l_${leftExprs.size}")()
+            leftExprs += a
+            a
+          }
+          
+          // Create an AttributeReference explicitly to avoid issues with unresolved aliases
+          val attributeRef = AttributeReference(alias.name, gpuExpr.dataType, 
+            gpuExpr.nullable, alias.metadata)(alias.exprId, alias.qualifier)
+          processed.put(GpuExpressionEquals(gpuExpr), attributeRef)
+          attributeRef
+      }
+    } else {
+      // This expression can be converted to AST
+      // Recursively process children, then convert this node to GPU
+      val convertedChildren = expr.childExprs.map { child =>
+        extractAndConvert(child, leftExprIds, rightExprIds, leftExprs, rightExprs, processed)
+      }
+      
+      // Convert to GPU and replace children with the processed versions
+      val gpuExpr = expr.convertToGpu()
+      if (convertedChildren.isEmpty) {
+        gpuExpr
       } else {
-        splitNonAstInternal(child, childAtt, left, right, substitutionMap, isLeft)
+        gpuExpr.withNewChildren(convertedChildren)
       }
     }
   }

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExecBase.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExecBase.scala
@@ -376,13 +376,17 @@ object GpuBroadcastNestedLoopJoinExecBase {
     }
   }
 
-  // A degenerate left-outer join means either the build or stream side has no columns.
-  // There are degenerate cases already found for the left-outer join with BuildRight.
-  // So now it just takes care of the left-outer join here.
+  // A degenerate outer join means either the build or stream side has no columns.
+  // This handles degenerate cases for left-outer join with BuildRight and
+  // right-outer join with BuildLeft (which are semantically equivalent - both
+  // keep all stream rows and add nulls from build when no match).
   private def isDegenerateLeftOuterJoin(joinType: JoinType, side: GpuBuildSide,
       builtBatch: LazySpillableColumnarBatch, streamAtts: Seq[Attribute]): Boolean = {
     joinType match {
-      case LeftOuter if side == GpuBuildRight =>  // now only support BuildRight
+      case LeftOuter if side == GpuBuildRight =>
+        // build or stream has no columns
+        builtBatch.numCols == 0 || streamAtts.isEmpty
+      case RightOuter if side == GpuBuildLeft =>
         // build or stream has no columns
         builtBatch.numCols == 0 || streamAtts.isEmpty
       case _ => false
@@ -394,7 +398,8 @@ object GpuBroadcastNestedLoopJoinExecBase {
       streamAttrs: Seq[Attribute],
       builtBatch: LazySpillableColumnarBatch,
       boundCondition: GpuExpression): Iterator[ColumnarBatch] = {
-    // Now only support BuildRight so the stream is left side.
+    // Supports both LeftOuter+BuildRight and RightOuter+BuildLeft.
+    // In both cases, all stream rows are kept with nulls from build when no match.
     if (streamAttrs.isEmpty) { // Stream has no columns
       new Iterator[ColumnarBatch] with TaskAutoCloseableResource {
         override def hasNext: Boolean = stream.hasNext

--- a/sql-plugin/src/main/spark320/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExec.scala
+++ b/sql-plugin/src/main/spark320/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExec.scala
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023-2024, NVIDIA CORPORATION.
+ * Copyright (c) 2023-2025, NVIDIA CORPORATION.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -61,7 +61,7 @@ class GpuBroadcastNestedLoopJoinMeta(
     if (isAstCondition) {
       // Try to extract non-ast-able conditions from join conditions
       val (remains, leftExpr, rightExpr) = AstUtil.extractNonAstFromJoinCond(
-        conditionMeta, left.output, right.output, true)
+        conditionMeta, left.output, right.output)
 
       // Reconstruct the childern with wrapped project node if needed.
       val leftChild =

--- a/sql-plugin/src/main/spark330db/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExec.scala
+++ b/sql-plugin/src/main/spark330db/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExec.scala
@@ -60,8 +60,8 @@ class GpuBroadcastNestedLoopJoinMeta(
 
     if (isAstCondition) {
       // Try to extract non-ast-able conditions from join conditions
-      val (remains, leftExpr, rightExpr) = AstUtil.extractNonAstFromJoinCond(conditionMeta,
-        left.output, right.output, true)
+      val (remains, leftExpr, rightExpr) = AstUtil.extractNonAstFromJoinCond(
+        conditionMeta, left.output, right.output)
 
       // Reconstruct the child with wrapped project node if needed.
       val leftChild =

--- a/sql-plugin/src/main/spark340/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExec.scala
+++ b/sql-plugin/src/main/spark340/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExec.scala
@@ -62,8 +62,8 @@ class GpuBroadcastNestedLoopJoinMeta(
 
     if (isAstCondition) {
       // Try to extract non-ast-able conditions from join conditions
-      val (remains, leftExpr, rightExpr) = AstUtil.extractNonAstFromJoinCond(conditionMeta,
-        left.output, right.output, true)
+      val (remains, leftExpr, rightExpr) = AstUtil.extractNonAstFromJoinCond(
+        conditionMeta, left.output, right.output)
 
       // Reconstruct the child with wrapped project node if needed.
       val leftChild =

--- a/tests/src/test/scala/com/nvidia/spark/rapids/AstUtilSuite.scala
+++ b/tests/src/test/scala/com/nvidia/spark/rapids/AstUtilSuite.scala
@@ -20,7 +20,7 @@ import org.mockito.Mockito.{mock, when}
 
 import org.apache.spark.sql.catalyst.expressions.{AttributeReference, AttributeSet, Expression}
 import org.apache.spark.sql.rapids.{GpuGreaterThan, GpuLength, GpuStringTrim}
-import org.apache.spark.sql.types.StringType
+import org.apache.spark.sql.types.{BooleanType, IntegerType, StringType}
 
 
 class AstUtilSuite extends GpuUnitTests {
@@ -187,21 +187,11 @@ class AstUtilSuite extends GpuUnitTests {
     rootMeta
   }
 
-  test("Non-Ast-able tree should not split"){
-    val l1 = AttributeReference("l1", StringType)()
-    val l2 = AttributeReference("l2", StringType)()
-    val (e, l, r) =
-      AstUtil.extractNonAstFromJoinCond(Some(buildTree1(l1, false)), Seq(l1), Seq(l2), false)
-    assertResult(true)(l.isEmpty)
-    assertResult(true)(r.isEmpty)
-    assertResult(true)(e.get.isInstanceOf[GpuStringTrim])
-  }
-
   test("Tree of single ast-able node should not split") {
     val l1 = AttributeReference("l1", StringType)()
     val l2 = AttributeReference("l2", StringType)()
     val (e, l, r) =
-      AstUtil.extractNonAstFromJoinCond(Some(buildTree1(l1, true)), Seq(l1), Seq(l2), false)
+      AstUtil.extractNonAstFromJoinCond(Some(buildTree1(l1, true)), Seq(l1), Seq(l2))
     assertResult(true)(l.isEmpty)
     assertResult(true)(r.isEmpty)
     assertResult(true)(e.get.isInstanceOf[GpuStringTrim])
@@ -211,7 +201,7 @@ class AstUtilSuite extends GpuUnitTests {
     val l1 = AttributeReference("l1", StringType)()
     val l2 = AttributeReference("l2", StringType)()
     val (e, l, r) =
-      AstUtil.extractNonAstFromJoinCond(Some(buildTree3(l1, l1, false)), Seq(l1), Seq(l2), false)
+      AstUtil.extractNonAstFromJoinCond(Some(buildTree3(l1, l1, false)), Seq(l1), Seq(l2))
     assertResult(true)(l.size == 1)
     assertResult(true)(l.exists(checkEquals(_, GpuLength(GpuStringTrim(l1)))))
     assertResult(true)(r.isEmpty)
@@ -232,7 +222,7 @@ class AstUtilSuite extends GpuUnitTests {
     val l1 = AttributeReference("l1", StringType)()
     val l2 = AttributeReference("l2", StringType)()
     val (e, l, r) =
-      AstUtil.extractNonAstFromJoinCond(Some(buildTree3(l1, l2, false)), Seq(l1), Seq(l2), false)
+      AstUtil.extractNonAstFromJoinCond(Some(buildTree3(l1, l2, false)), Seq(l1), Seq(l2))
     assertResult(true)(l.size == 1)
     assertResult(true)(l.exists(checkEquals(_, GpuLength(GpuStringTrim(l1)))))
     assertResult(true)(r.size == 1)
@@ -245,10 +235,72 @@ class AstUtilSuite extends GpuUnitTests {
     val l1 = AttributeReference("l1", StringType)()
     val l2 = AttributeReference("l2", StringType)()
     val (e, l, r) =
-      AstUtil.extractNonAstFromJoinCond(Some(buildTree3(l1, l2, true)), Seq(l1), Seq(l2), false)
+      AstUtil.extractNonAstFromJoinCond(Some(buildTree3(l1, l2, true)), Seq(l1), Seq(l2))
     assertResult(true)(l.size == 0)
     assertResult(true)(r.size == 0)
     assertResult(true)(checkEquals(e.get,
       GpuGreaterThan(GpuLength(GpuStringTrim(l1)), GpuLength(GpuStringTrim(l2)))))
+  }
+
+  // Build a simple tree representing: cast(a:int as boolean)
+  // This simulates a condition like: SELECT * FROM t1 JOIN t2 ON t1.intCol
+  // where intCol is cast to boolean for the join condition
+  private[this] def buildTreeWithCast(attr: AttributeReference): BaseExprMeta[Expression] = {
+    // Create the mock cast expression (int -> boolean)
+    val castExpr = mock(classOf[Expression])
+    when(castExpr.references).thenReturn(AttributeSet(Seq(attr)))
+    when(castExpr.dataType).thenReturn(BooleanType)
+    
+    val castMeta = mock(classOf[BaseExprMeta[Expression]])
+    when(castMeta.childExprs).thenReturn(Seq.empty)
+    when(castMeta.canSelfBeAst).thenReturn(false) // Cast cannot be AST
+    when(castMeta.convertToGpu).thenReturn(castExpr)
+    when(castMeta.wrapped).thenReturn(castExpr)
+    castMeta
+  }
+
+  test("Top-level non-AST expression on single side should be extracted") {
+    val l1 = AttributeReference("l1", IntegerType)()
+    val r1 = AttributeReference("r1", IntegerType)()
+
+    // Test with left side only
+    val (expr, leftExprs, rightExprs) =
+      AstUtil.extractNonAstFromJoinCond(Some(buildTreeWithCast(l1)), Seq(l1), Seq(r1))
+
+    // Should extract the cast to the left side
+    assertResult(1)(leftExprs.size)
+    assertResult(0)(rightExprs.size)
+    assertResult(true)(expr.isDefined)
+
+     // The returned expression should be an attribute reference (the replacement)
+     assertResult(true)(expr.get.isInstanceOf[AttributeReference])
+ 
+     // The attribute should match the one from the left expression alias
+     val leftAlias = leftExprs.head.asInstanceOf[GpuAlias]
+     val expectedAttr = AttributeReference(leftAlias.name, leftAlias.child.dataType,
+       leftAlias.child.nullable, leftAlias.metadata)(leftAlias.exprId, leftAlias.qualifier)
+     assertResult(expectedAttr)(expr.get)
+  }
+
+  test("Top-level non-AST expression on right side should be extracted to right") {
+    val l1 = AttributeReference("l1", IntegerType)()
+    val r1 = AttributeReference("r1", IntegerType)()
+    
+    // Test with right side only
+    val (expr, leftExprs, rightExprs) =
+      AstUtil.extractNonAstFromJoinCond(Some(buildTreeWithCast(r1)), Seq(l1), Seq(r1))
+    
+    // Should extract the cast to the right side
+    assertResult(0)(leftExprs.size)
+    assertResult(1)(rightExprs.size)
+    
+     // The returned expression should be an attribute reference (the replacement)
+     assertResult(true)(expr.get.isInstanceOf[AttributeReference])
+     
+     // The attribute should match the one from the right expression alias
+     val rightAlias = rightExprs.head.asInstanceOf[GpuAlias]
+     val expectedAttr = AttributeReference(rightAlias.name, rightAlias.child.dataType,
+       rightAlias.child.nullable, rightAlias.metadata)(rightAlias.exprId, rightAlias.qualifier)
+     assertResult(expectedAttr)(expr.get)
   }
 }

